File: 3123-find-edges-in-shortest-paths/3123-find-edges-in-shortest-paths.java
class Solution {
    public boolean[] findAnswer(int n, int[][] edges) {
        
        ArrayList<ArrayList<Pair>> adj = new ArrayList<>();
        
        for(int i = 0 ; i < n ; i++){
            adj.add(new ArrayList<>());
        }
        
        for(int ed[] : edges){
            int u = ed[0];
            int v = ed[1];
            int w = ed[2];
            
            adj.get(u).add(new Pair(v,w));
            adj.get(v).add(new Pair(u,w));
        }
        
        int dist[] = new int[n];
        Arrays.fill(dist,Integer.MAX_VALUE);
        dist[0] = 0;
        
        ArrayList<Integer>[] parent = new ArrayList[n];
        
        PriorityQueue<Pair> q = new PriorityQueue<>((a,b)->{return a.wt-b.wt;});
        q.add(new Pair(0,0));
        
        while(!q.isEmpty()){
            
            Pair pop = q.poll();
            int node = pop.node;
            int wt = pop.wt;
            
            for(Pair nei : adj.get(node)){
                
                int adjNode = nei.node;
                int adjWt = nei.wt;
                
                int adjDist = wt+adjWt;
                
                if(dist[adjNode] < adjDist){
                    continue;
                }
                
                if(adjDist == dist[adjNode]){
                    parent[adjNode].add(node);
                }else{
                    parent[adjNode] = new ArrayList<>();
                    parent[adjNode].add(node);
                    dist[adjNode] = adjDist;
                    q.add(new Pair(adjNode,adjDist));
                }
            }
        }
        
        Queue<Integer> que = new LinkedList<>();
        que.add(n-1);
        
        boolean ans[] = new boolean[edges.length];
        
        HashMap<String,Integer> edgeInds = new HashMap<>();
        for(int i = 0 ; i < edges.length ; i++){
            int ed[] = edges[i];
            int u = ed[0];
            int v = ed[1];
            
            edgeInds.put(u+","+v,i);
            edgeInds.put(v+","+u,i);
        }
        
        HashSet<Integer> vis = new HashSet<>();
        
        while(!que.isEmpty()){
            
            int pop = que.poll();
            
            if(parent[pop] == null){
                continue;
            }
            
            for(int nei : parent[pop]){
                
                String edg = pop+","+nei;
                ans[edgeInds.get(edg)] = true;
                
                if(!vis.contains(nei)){
                    vis.add(nei);
                    que.add(nei);
                }
            }
        }
        
        return ans;
    }
    class Pair{
        int node;
        int wt;
        Pair(int node, int wt){
            this.node = node;
            this.wt = wt;
        }
    }
}

File: 3123-find-edges-in-shortest-paths/3123-find-edges-in-shortest-paths.cpp
#define pii pair<int, int>
#define F first
#define S second

typedef long long int ll;

class Solution {
    vector<vector<pii>> g;
    vector<vector<int>> parents;
    
public:
    vector<bool> findAnswer(int n, vector<vector<int>>& edges) {
        g.clear(), g.resize(n);
        parents.clear(), parents.resize(n);
        
        for (auto e: edges) {
            int u = e[0], v = e[1], w = e[2];
            g[u].push_back({v, w});
            g[v].push_back({u, w});
        }
        
        vector<ll> dist(n, 1e18);
        dist[0] = 0;
        
        set<pii> s;
        for (int j = 0; j < n; j ++) s.insert({dist[j], j});
        
        while (!s.empty()) {
            int cur = s.begin() -> S;
            s.erase(s.begin());
            
            for (auto [v, w]: g[cur]) {
                ll can = dist[cur] + w;
                if (can > dist[v]) continue;
                
                if (can == dist[v]) parents[v].push_back(cur);
                else {
                    parents[v] = {cur};
                    
                    s.erase({dist[v], v});
                    dist[v] = can;
                    s.insert({dist[v], v});
                }
            }
        }
        
        map<pii, int> edges_ind;
        for (int i = 0; i < edges.size(); i ++) {
            edges_ind[{edges[i][0], edges[i][1]}] = i; 
            edges_ind[{edges[i][1], edges[i][0]}] = i; 
        }
        
        vector<bool> ans(edges.size(), false);
        vector<bool> vis(n, false);
        
        queue<int> q;
        q.push(n-1);        
        while (!q.empty()) {
            int cur = q.front();
            q.pop();
            
            for (auto nxt: parents[cur]) {
                // cout << cur << " " << nxt << " " << edges_ind[{cur, nxt}] << endl;
                ans[edges_ind[{cur, nxt}]] = true;
                if (!vis[nxt]) {
                    q.push(nxt);
                    vis[nxt] = true;
                }
            }
        }
        return ans;
    }
};